lib/src/command/add.dart

```diff
@@ -0,0 +1,237 @@
+      dependencies: dependencies,
+      devDependencies: devDependencies,
+      dependencyOverrides: original.dependencyOverrides.values,
+    );
+  }
+
+  /// Parse [PackageInfo] from [package].
+  ///
+  /// [package] must be written in the format
+  /// `<package-name>[:<version-constraint>]`, where quotations should be used
+  /// if necessary.
+  ///
+  /// Examples:
+  /// ```bash
+  /// retry
+  /// retry:2.0.0
+  /// retry:^2.0.0
+  /// retry:'>=2.0.0'
+  /// retry:'>2.0.0 <3.0.1'
+  /// 'retry:>2.0.0 <3.0.1'
+  /// retry:any
+  /// ```
+  ///
+  /// If a version constraint is provided when the `--path` or any of the
+  /// `--git-<option>` options are used, a [PackageParseError] will be thrown.
+  ///
+  /// If both `--path` and any of the `--git-<option>` options are defined,
+  /// a [PackageParseError] will be thrown.
+  ///
+  /// If any of the other git options are defined when `--git-url` is not
+  /// defined, an error will be thrown.
+  PackageInfo _parsePackage(String package) {
+    ArgumentError.checkNotNull(package, 'packages');
+
+    var git = <String, String>{};
+    if (gitUrl != null) git['url'] = gitUrl;
+    if (gitRef != null) git['ref'] = gitRef;
+    if (gitPath != null) git['path'] = gitPath;
+    if (git.isEmpty) git = null;
+
+    Map<String, String> hostInfo;
+    if (hostUrl != null) hostInfo = {'url': hostUrl};
+
+    PackageInfo parsedPackage;
+
+    try {
+      parsedPackage = PackageInfo.from(package,
+          path: path,
+          git: git,
+          pubspecPath: entrypoint.pubspecPath,
+          hostInfo: hostInfo);
+    } on PackageParseException catch (exception) {
+      usageException(exception.message);
+    }
+
+    return parsedPackage;
+  }
+
+  /// Retrieves the result of version resolution on [package], and assert that
+  /// it is within the original user's expectations, throwing a [DataError]
+  /// otherwise.
+  Future<PackageId> _getAndAssertResultPackage(
+      SolveResult result, PackageInfo package) async {
+    ArgumentError.checkNotNull(result, 'result');
+    ArgumentError.checkNotNull(package, 'package');
+
+    /// The `orElse` scenario should not happen because it would have been
+    /// discovered in the version resolution process.
+    final resultPackage = result.packages
+        .firstWhere((packageId) => packageId.name == package.name);
+
+    if (package is HostedPackageInfo &&
+        package.constraint != null &&
+        !package.constraint.allows(resultPackage.version)) {
+      dataError('${package.name} resolved to ${resultPackage.version} which '
+          'does not match the input ${package.constraint}! Exiting.');
+    }
+
+    return resultPackage;
+  }
+
+  /// Writes the changes to the pubspec file
+  void _updatePubspec(PackageId resultPackage, PackageInfo package) {
+    ArgumentError.checkNotNull(resultPackage, 'resultPackage');
+    ArgumentError.checkNotNull(package, 'package');
+
+    if (entrypoint.pubspecPath == null) {
```

Comment:
the `entrypoint` getter will throw if there is not pubspec.yaml. You don't need this check.